File: Candy.java
import java.util.Arrays;

//Time Complexity=O(n)
//Space Complexity=O(n)
public class Candy {

    public int candy(int[] ratings) {

        if(ratings==null || ratings.length==0 ) return 0;
        int n=ratings.length;
        if(n==1) return 1;

        int[] result=new int[n];
        int sum;
        Arrays.fill(result,1);

        for(int i=1;i<n;i++){
            if(ratings[i]>ratings[i-1]){
                result[i]=result[i-1]+1;
            }
        }
        sum=result[n-1];

        for(int i=n-2;i>=0;i--){
            if(ratings[i]>ratings[i+1]){
                result[i]=Math.max(result[i],result[i+1]+1);
            }
            sum+=result[i];
        }

        return sum;
    }
}

File: candy.py
#Time complexity: O(n)
#Space complexity: O(n)

class Solution:
    def candy(self, ratings: List[int]) -> int:
        n = len(ratings)
        res = [1 for _ in range(n)]
        if ratings == None or n == 0:
            return 0
        if n == 1:
            return 1
        for i in range(1, n):
            if ratings[i-1] < ratings[i]:
                res[i] = 1 + res[i-1]
        sumR = res[n-1]
        for i in range(n-2, -1, -1):
            if ratings[i+1] < ratings[i]:
                res[i] = max(res[i], 1+res[i+1])
            sumR += res[i]
        return sumR
        
